### PR TITLE
[CVAT-M2] Oracles M2 API

### PR DESCRIPTION
**Pull request by zhiltsov-max**
## Description

Changes for the new version of Human APP.
Please, do not merge before other services are ready for this.

## Summary of changes

- Updated webhook endpoints in oracles
- Updated webhook events naming and contents
- Updated tests

Specification: https://docs.humanprotocol.org/hub/human-tech-docs/architecture/components/exchange-oracle/standard

## Related issues
[Keywords for linking issues](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests)

### CODE REVIEW

File: packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py [hunk 1]
```diff
@@ -131,13 +133,19 @@ def _selectable_fields(self) -> list[str] | None:
         )
         return list(selectable_fields_enum.__members__)
 
+    @classmethod
+    def get_default_field_value(cls, field_name: str) -> Any:
+        default_value = cls.model_fields[field_name].get_default()
+        if isinstance(default_value, fastapi.params.Query):
```
Comment (zhiltsov-max) [inline packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py:139]: Probably, it's better to check for `params.FieldInfo` instead for better compatibility with other possible field declarations.

**Comment by Marishka17** [comment]
@zhiltsov-max, Exchange oracle fixes:

  - Renamed `task_id` to `assignment_id` in the `POST /webhook` body when submission_rejected event occurs
  https://github.com/humanprotocol/human-protocol/blob/fdf65b03c631ad5179336d2f6f7eb11fc29beb79/packages/examples/cvat/exchange-oracle/src/core/oracle_events.py#L33
  commit - https://github.com/humanprotocol/human-protocol/pull/2094/commits/0cc7dfd825a387a6c9ef8b85daea08e0dfd40129
  
  - Return 401 instead of 403 when auth token is missing
  https://github.com/humanprotocol/human-protocol/blob/fdf65b03c631ad5179336d2f6f7eb11fc29beb79/packages/examples/cvat/exchange-oracle/src/endpoints/authentication.py#L16
  commit - https://github.com/humanprotocol/human-protocol/pull/2094/commits/23cfef69b892dc67923568ca2e224f10ae557088
  
  - Added description and examples for JobsFilter::fields that will be displayed on swagger doc
  https://github.com/humanprotocol/human-protocol/blob/fdf65b03c631ad5179336d2f6f7eb11fc29beb79/packages/examples/cvat/exchange-oracle/src/endpoints/exchange.py#L69
  commit - https://github.com/humanprotocol/human-protocol/pull/2094/commits/fdf65b03c631ad5179336d2f6f7eb11fc29beb79
  
  - Made GET /assignment query params optional
  https://github.com/humanprotocol/human-protocol/blob/fdf65b03c631ad5179336d2f6f7eb11fc29beb79/packages/examples/cvat/exchange-oracle/src/endpoints/exchange.py#L263
  
  - Moved `assignment_id` from query params into request body (POST /assignment/resign)
  https://github.com/humanprotocol/human-protocol/blob/fdf65b03c631ad5179336d2f6f7eb11fc29beb79/packages/examples/cvat/exchange-oracle/src/endpoints/exchange.py#L402
  commit - https://github.com/humanprotocol/human-protocol/pull/2094/commits/29d77c4177e03ef360ff548ec333be0f23771a3c
  
  - Fixed sorting 
  https://github.com/humanprotocol/human-protocol/blob/fdf65b03c631ad5179336d2f6f7eb11fc29beb79/packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py#L84
  commit - https://github.com/humanprotocol/human-protocol/pull/2094/commits/95adede95bc069c309d0f86ec844420834c0315a
  
  - Fixed pagination
  commit - https://github.com/humanprotocol/human-protocol/pull/2094/commits/ca09ae2bca9d57ba9fa1490f652589ed4e04a482
  
  - Added missing `updated_at` in Job response
  https://github.com/humanprotocol/human-protocol/blob/fdf65b03c631ad5179336d2f6f7eb11fc29beb79/packages/examples/cvat/exchange-oracle/src/endpoints/serializers.py#L55
  commit - https://github.com/humanprotocol/human-protocol/pull/2094/commits/d24cdb0fa29e8d3d7f49c45dd1a575f7f0f4ae95
  
  - Fixed missing `reward_token` in AssignmentResponse/JobResponse
  https://github.com/humanprotocol/human-protocol/blob/fdf65b03c631ad5179336d2f6f7eb11fc29beb79/packages/examples/cvat/exchange-oracle/src/endpoints/serializers.py#L145
  commit - https://github.com/humanprotocol/human-protocol/pull/2094/commits/29d77c4177e03ef360ff548ec333be0f23771a3c

File: packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py (hunk 1 above)
Comment (Marishka17) [reply packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py:139]:
```suggestion
        if isinstance(default_value, fastapi.params.FieldInfo):
```

Comment (Marishka17) [inline packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py:139]:
```suggestion
        if isinstance(default_value, fastapi.params.FieldInfo):
```

Comment (Marishka17) [reply packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py]: done

File: packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py [hunk 2]
```diff
@@ -1,100 +1,306 @@
...
+empty_result = {
+    "page": 0,
+    "total_pages": 0,
+    "total_results": 0,
+    "results": [],
+    "page_size": 5,  # default page size
+}
+
+
+def test_empty_list_jobs_200_with_address(client: TestClient) -> None:
```
Comment (zhiltsov-max) [inline packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py:60]: Please rename tests to use pattern "test_can_*". It makes it more clear what is the expected outcome in the case being tested.

File: packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py [hunk 3]
```diff
@@ -1,100 +1,306 @@
...
+    assert (
+        response.text
+        == '{"errors":[{"field":"page_size","message":"Input should be less than or equal to 10"}]}'
+    )
+
+
+def test_list_jobs_200_with_address_and_pagination(client: TestClient) -> None:
+    def validate_result(
+        response: Response,
+        result_to_compare_with: list[str],
```
Comment (zhiltsov-max) [inline packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py:95]: Typically it's called `expected`, I guess.

File: packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py [hunk 4]
```diff
@@ -1,100 +1,306 @@
...
+    response = client.get("/job", headers=get_auth_header(), params={"page_size": 1000})
 
-        user = User(
-            wallet_address=user_address,
-            cvat_email="test@hmt.ai",
-            cvat_id=1,
+    assert response.status_code == 400
+    assert (
+        response.text
+        == '{"errors":[{"field":"page_size","message":"Input should be less than or equal to 10"}]}'
```
Comment (zhiltsov-max) [inline packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py:88]: It'better to take the limit from the config or mock it here.

File: packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py [hunk 5]
```diff
@@ -1,100 +1,306 @@
...
+        assignment = Assignment(
+            id=str(uuid.uuid4()),
+            user_wallet_address=user_address,
+            cvat_job_id=job.cvat_id,
+            expires_at=datetime.now() + timedelta(days=1),
+        )
+        session.add(assignment)
+        job.status = JobStatuses.in_progress
+        session.commit()
+        sleep(0.5)
```
Comment (zhiltsov-max) [inline packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py:267]: Why this sleep is needed?

File: packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py [hunk 6]
```diff
@@ -252,89 +416,536 @@ def test_register_400(client: TestClient) -> None:
...
+        open("tests/utils/manifest.json") as data,
+        patch("src.endpoints.serializers.get_escrow_manifest") as mock_get_manifest,
+    ):
+        manifest = json.load(data)
+        mock_get_manifest.return_value = manifest
+
+        for sort_field in (
+            "chain_id",
+            "job_type",
+            # TODO: "reward_amount",
```
Comment (zhiltsov-max) [inline packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py:645]:
```suggestion
```

File: packages/examples/cvat/recording-oracle/src/handlers/validation.py [hunk 7]
```diff
@@ -167,41 +171,52 @@ def _handle_validation_result(self, validation_result: ValidationResult):
...
+                                self.manifest.validation.min_quality,
+                            ),
+                        )
+                        for rejected_job_id, reason in validation_result.rejected_jobs.items()
+                        if not isinstance(reason, TooFewGtError)
                     ]
                 ),
             )
+        else:
+            raise TypeError(f"Unexpected validation result {type(validation_result)=}")
```
Comment (zhiltsov-max) [inline packages/examples/cvat/recording-oracle/src/handlers/validation.py:219]:
```suggestion
            raise TypeError(f"Unexpected validation result {type(validation_result)}")
```

File: packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py (hunk 5 above)
Comment (Marishka17) [reply packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py]: This test should also check that `updated_at` will correspond to the latest updated job. Therefore I emulate sequential job updates by calling `session.commit()` after each change. Using `sleep` is optional, deleted that.

File: packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py (hunk 2 above)
Comment (Marishka17) [reply packages/examples/cvat/exchange-oracle/tests/api/test_exchange_api.py]: Done

**Comment by zhiltsov-max** [comment]
Need to fix invalid `updated_at` filter in `/job` - it doesn't take into account children updates, while serializer does. Probably, it's better to update the parent objects on children updates instead of making complex DB queries in the `list` requests.

File: packages/examples/cvat/recording-oracle/src/handlers/process_intermediate_results.py [hunk 8]
```diff
@@ -1093,8 +1094,8 @@ def process_intermediate_results(  # noqa: PLR0912
 
 
 def parse_annotation_metafile(metafile: io.RawIOBase) -> AnnotationMeta:
-    return AnnotationMeta.parse_raw(metafile.read())
+    return AnnotationMeta.model_validate(metafile.read())
```
Comment (zhiltsov-max) [inline packages/examples/cvat/recording-oracle/src/handlers/process_intermediate_results.py:1097]:
```suggestion
    return AnnotationMeta.model_validate_json(metafile.read())
```

File: packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+from fastapi_filter.contrib.sqlalchemy import Filter as _Filter
+from pydantic import BaseModel, Field, ValidationInfo, field_validator
+from pydantic.fields import FieldInfo
+
+from src.utils.enums import BetterEnumMeta
+
+
+class OrderingDirection(str, Enum, metaclass=BetterEnumMeta):
+    asc = "asc"
+    desc = "desc"
```
Comment (zhiltsov-max) [inline packages/examples/cvat/exchange-oracle/src/endpoints/filtering.py:18]:
```suggestion
    asc = "ASC"
    desc = "DESC"
```

**Comment by Dzeranov** [comment]
Please, remove .vscode folder.

File: packages/examples/cvat/exchange-oracle/src/crons/cvat/state_trackers.py [hunk 10]
```diff
@@ -118,6 +123,16 @@ def track_assignments(logger: logging.Logger) -> None:
...
 
+        jobs_to_be_updated = [a.job for a in assignments]
+        tasks_to_be_updated = [j.task for j in jobs_to_be_updated]
+        projects_to_be_updated = [t.project for t in tasks_to_be_updated]
+        cvat_service.touch_jobs(session, {job.id for job in jobs_to_be_updated})
+        cvat_service.touch_tasks(session, {task.id for task in tasks_to_be_updated})
+        cvat_service.touch_projects(session, {project.id for project in projects_to_be_updated})
+        del jobs_to_be_updated
+        del tasks_to_be_updated
+        del projects_to_be_updated
```
[the remark this answers is not in the document]
Comment (Bobronium) [reply packages/examples/cvat/exchange-oracle/src/crons/cvat/state_trackers.py]: Removing the references to the objects so they could be destroyed by garbage collector earlier, I guess.

File: packages/examples/cvat/exchange-oracle/src/crons/cvat/state_trackers.py [hunk 11]
```diff
@@ -118,6 +123,16 @@ def track_assignments(logger: logging.Logger) -> None:
 
             cvat_service.expire_assignment(session, assignment.id)
 
+        jobs_to_be_updated = [a.job for a in assignments]
+        tasks_to_be_updated = [j.task for j in jobs_to_be_updated]
+        projects_to_be_updated = [t.project for t in tasks_to_be_updated]
+        cvat_service.touch_jobs(session, {job.id for job in jobs_to_be_updated})
+        cvat_service.touch_tasks(session, {task.id for task in tasks_to_be_updated})
+        cvat_service.touch_projects(session, {project.id for project in projects_to_be_updated})
```
[the remark this answers is not in the document]
Comment (Bobronium) [reply packages/examples/cvat/exchange-oracle/src/crons/cvat/state_trackers.py]: Done!

File: packages/examples/cvat/exchange-oracle/src/models/cvat.py [hunk 12]
```diff
@@ -65,6 +65,9 @@ class Project(Base):
     def __repr__(self) -> str:
         return f"Project. id={self.id}"
 
+    def touch(self, session: Session) -> None:
+        session.query(Project).filter(Project.id == self.id).update({Project.updated_at: utcnow()})
+
```
[the remark this answers is not in the document]
Comment (Bobronium) [reply packages/examples/cvat/exchange-oracle/src/models/cvat.py]: Replaced this with `src.services.cvat.touch`

File: packages/examples/cvat/exchange-oracle/src/crons/cvat/state_trackers.py [hunk 13]
```diff
@@ -213,6 +236,7 @@ def track_task_creation(logger: logging.Logger, session: Session) -> None:
                     )
 
                 completed.append(upload)
+                upload.task.touch(session, touch_parent=True)
```
[the remark this answers is not in the document]
Comment (Bobronium) [reply packages/examples/cvat/exchange-oracle/src/crons/cvat/state_trackers.py]: Done!

**Review by Dzeranov — APPROVED** [review]

**Pull request merged**